### ISSUE TITLE
Allow variables to be looked up lazily - switch VariablesMap to sort of VariablesResolver.

### ISSUE THREAD

**Issue by OndraZizka** [thymeleaf/thymeleaf#101]
When extending ThymeLeaf's IContext, the variables map to be created can become really huge, and very inefficient to create upfront. Think of CDI-backed values. Or database-backed.

It would be nice if ThymeLeaf allowed lazy lookup, possibly using some kind of cached map.

**Comment by melix** [comment]
I second that issue. I wrote a PR to integrate Thymeleaf into JBake (see jonbullock/jbake#91) and one major issue is that we can't have lazy items like we do with Freemarker or Groovy templates. Even if I create a custom implementation of `VariableMap`, there is code in `AbstractContext` (and other classes) which copies the contents of the `CustomMap` into a new `VariableMap` or `HashMap`, so not the same implementation unfortunately.

This means that the _whole_ content of the document database need to be present in the context while rendering. This might be sufficient for small sites, definitely not for sites will lots of pages.

**Comment by christianbauer** [comment]
Yes, this isn't a great design choice. A simple Object resolve(String name) interface must be enough for the EL resolver. And then there is this gem which really makes any attempt at a workaround futile:

```
private Object createEvaluationRoot() {

    final VariablesMap<String, Object> contextVariables = this.context.getVariables();
    // We create a new VariablesMap instance using its constructor instead of cloning the existing one
    // because we want to avoid undesirable interactions like, for example, those that could happen
    // if we executed putAll on a WebVariablesMap object (which would add those variables to the HttpServletRequest
    // and therefore make them available to the whole page and not just the local variable scope).
    final VariablesMap<String,Object> newEvaluationRoot = new VariablesMap<String, Object>(contextVariables);
```

**Comment by danielfernandez** [comment]
Just a note: possibility for 3.0: Create a `ThymeleafLazyLoader` interface so that, if an executed expression is resolved to an object implementing it, its `load()` method is called in order to obtain the real data that should be considered the result of the expression. This `load()` should only be called once per template rendering.

This should solve the scenario for data that might never be needed in the template (because of `th:if` conditions, for example).

The scenario of very large data sets that we prefer not to be completely loaded in memory will be solved in 3.0 with the new architecture (which will not need the entire page DOM to be generated before rendering) and at the data side with the already-existing support for iterating `Iterator` objects with e.g. `th:each`. So allowing the developers to create lazy loaders returning a properly built `Iterator` on e.g. a `ResultSet` should be enough.

**Comment by danielfernandez** [comment]
Note for self: evaluate `javax.inject.Provider<T>` in JSR-330 as a possible interface (maybe along with a Thymeleaf-specific one) being implemented by objects to be lazy-loaded into the variables map.

Also, note that according to the new v3 architecture, it should be the `IVariablesMap` implementations (web and non-web) the ones performing these lazy-loading resolutions. Also, we should make sure it is possible for the users to create their own implementations of `IProcessingContext` returning a custom-made implementation of `IVariablesMap` (even if the new local variables support would be somewhat hard to code, give them at least the opportunity to) in order to support scenarios like the JBake one --- even if that specific scenario would be supported out-of-the-box, but just in case anyone else needs to do any other complex operations with variable resolution…

**Comment by danielfernandez** [comment]
This has been implemented in Thymeleaf 3.0 by means of the `org.thymeleaf.context.ILazyContextVariable` interface and its default abstract implementation `org.thymeleaf.context.LazyContextVariable`.

Context variables that implement the `ILazyContextVariable` interface will be resolved in the moment of being executed, effectively _lazy-loading_ their value. For example:

``` java
context.setVariable(
     "users",
     new LazyContextVariable<List<User>>() {
         @Override
         protected List<User> loadValue() {
             return databaseRepository.findAllUsers();
         }
     });
```

This variable can be used without noticing its _lazyness_, in code such as:

``` html
<ul>
  <li th:each="u : ${users}" th:text="${u.name}">user name</li>
</ul>
```

But at the same time, will never be initialized (its `loadValue()` method will never be called) if `condition` evaluates to `false` in code such as:

``` html
<ul th:if="${condition}">
  <li th:each="u : ${users}" th:text="${u.name}">user name</li>
</ul>
```

Some tests exemplifying its use can be found here: https://github.com/thymeleaf/thymeleaf-tests/tree/3.0-spring4/src/test/resources/templateengine/features/lazy

**Comment by danielfernandez** [comment]
This is already in `3.0.0-SNAPSHOT`
